### PR TITLE
W-10003846 Error Notifier SOQL Query Limits check

### LINKED ISSUE #1881: Error Notifier: No LIMIT on Error SOQL Query Can Result In 50,000 Row Limit or Heap Size Errors
* **Do you want to request a *feature* or report a *bug*?**
Bug

* **What is the current behavior?**
Customers with large volume of Error records are at risk of heap size errors or the 50,000 row limit error when the EDA Error Notifier runs.

StackTrace from Heap Size Error:
```Developer script exception from {Customer} : 'hed.ERR_Notifier' for job id '707xxxxxxxxxxxx' : Apex heap size too large: 13498444
ExtendedMessage: Apex script unhandled exception by user/organization: 005xxxxxxxxxxxx/00Dxxxxxxxxxxxx
Failed to invoke future method 'private static void sendErrorQueueEmailNotificationFuture(List<String>)' on class 'hed.ERR_Notifier' for job id '707xxxxxxxxxxxx'
caused by: System.LimitException: Apex heap size too large: 13498444
External entry point
Class.hed.ERR_Notifier.sendErrorQueueEmailNotification: line 130, column 1
Class.hed.ERR_Notifier.sendErrorQueueEmailNotificationFuture: line 115, column 1
```

### Root Cause

When ERR_Notifier's sendErrorQueueEmailNotification() method is called, we run a SOQL query to gather all Error__c records where Email_Sent__c = false. If this is > 50,000 records, an error is thrown (again) and the notification isn't actually sent. Likewise, if a customer has thousands of Error records not yet sent and their Full_Message__c field is lengthy, this can result in heap size errors.

### Steps to Repeat in a Scratch Org:

- Ensure Education Cloud Settings are set to email All Sys Admins.
- Create 50,001 Error__c records in an EDA org.
- Verify that when the Error Notifier runs, there is a 50K row error for the job in Setup --> Apex Jobs.

### Workaround

Typically, we would suggest that the customer fixes the issue(s) that caused all of the Error records to be created in the first place. And then start either deleting some of the older Error records or mark Email_Sent__c to TRUE so that they are excluded from the query going forward.

### Product Recommendations
Place a limit on records being pulled into ERR_Notifier (total records and/or time-based).

### PR DESCRIPTION
**Pull request by iskander-sfdc**
Added SOQL limit (500 errors), error age (48 hours) and apex heap limit checks when sending errors via email and chatter

# Critical Changes

# Changes

# Issues Closed
fixes #1881 

# New Metadata

## Unpackaged Metadata

# Deleted Metadata

# Testing Notes
https://gus.lightning.force.com/lightning/r/0D5EE00000Tp16F0AR/view

### CODE REVIEW

File: force-app/main/default/classes/ERR_Notifier_TEST.cls [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+     * only retrieve records created in the last 48 hours.
+     */
+    @IsTest
+    private static void shouldOnlyRetrieveRecentErrorLogEntries() {
+        ERR_Notifier.MAX_AGE_FOR_ERRORS = Datetime.now().addDays(-3);
+        ERR_Notifier.MAX_RECORDS = 10;
+
+        List<Error__c> errors = new List<Error__c>();
+        // The oldest error will be 80 days old
+        Datetime dt = Datetime.now().addDays(80).addHours(-1);
```
Comment (kg345) [inline force-app/main/default/classes/ERR_Notifier_TEST.cls:114]: Is `addHours(-1)` necessary?

File: force-app/main/default/classes/ERR_Notifier_TEST.cls [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+        System.assertEquals(expectedEmails.size(), recipientEmails.size(), 'Number of emails should match');
+        System.assert(expectedEmails.containsAll(recipientEmails), 'Expected and retrieved emails should be the same');
+    }
+
+    /**********************************************************************************************
+     * @description Confirm that the method to retrieve Error log record pending notification
+     * only retrieve records created in the last 48 hours.
+     */
+    @IsTest
+    private static void shouldOnlyRetrieveRecentErrorLogEntries() {
```
Comment (kg345) [inline force-app/main/default/classes/ERR_Notifier_TEST.cls:108]: Would it be worth splitting this into two separate tests: one to assert against max age (where it doesn't reach max records) and the other to assert max records (which this one is doing)? For this one, it isn't necessarily testing max age as the max records gets hit first.

File: force-app/main/default/classes/ERR_Notifier_TEST.cls [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+        for (Error__c err : getErrors()) {
+            System.assertEquals(
+                false,
+                err.Email_Sent__c,
+                'All errors pending notification should NOT have been emailed because over heap size limit'
+            );
+        }
+    }
+
+    // Helpers
```
Comment (kg345) [inline force-app/main/default/classes/ERR_Notifier_TEST.cls:288]: Do we typically add test data methods to `UTIL_UnitTestData_TEST` or is it preferred within the test class?

File: force-app/main/default/classes/ERR_Notifier.cls [hunk 4]
```diff
@@ -28,162 +28,255 @@
...
+            sendList.addAll(this.getSystemAdminEmails());
+        }
+        return sendList;
+    }
 
+    /**********************************************************************************************
+     * @description Get the active System Administrators' emails.
+     * @return List<String> Emails
+     */
+    @TestVisible
```
Comment (kg345) [inline force-app/main/default/classes/ERR_Notifier.cls:130]: I didn't see this referenced in the test class, is this annotation needed?

**Review by kg345 — APPROVED** [review]
Looks good to me! Confirmed following the reproduction steps and using snowfakery that the query limit error no longer exists. I left some comments asking some questions and minor things, but nothing critical. 

I'll work on creating some test scenarios related to this as I did not find any scenarios outside of [this one testing the old settings](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001Kn6IAE/view?ws=%2Flightning%2Fr%2FADM_Test_Suite__c%2Fa81B00000004eqVIAQ%2Fview)

File: force-app/main/default/classes/ERR_Notifier.cls [hunk 5]
```diff
@@ -28,162 +28,255 @@
...
     public static NotificationOptions notificationOptions = new NotificationOptions();
-    
+
+    /** @description Filtering options to reduce the number of records processed during the notify job */
+    @TestVisible
+    private static Integer MAX_RECORDS = 500;
+    @TestVisible
+    private static Datetime MAX_AGE_FOR_ERRORS = System.now().addHours(-48);
+
+    public Boolean sendEmailAsync { get; set; }
```
[the remark this answers is not in the document]
Comment (iskander-sfdc) [reply force-app/main/default/classes/ERR_Notifier.cls:46]: EDA sends error notifications in async mode atm. In unit tests we need to send them in sync mode so we could tweak the heap limits. I added this parameter so the service could send emails in sync and async mode when needed (in tests and when needed going forward). The default is true - async mode.

File: force-app/main/default/classes/ERR_Notifier.cls [hunk 6]
```diff
@@ -28,162 +28,255 @@
...
+     * @description Sends error email notifications in a future (asynchronously). It will send an email with all the
+     *              existing error records not already sent in a notification.
+     * @param context The context that triggered the error notification.
+     * @param sendList The list of email addresses to send notifications to.
+     * @return void
+     */
     @future
     private static void sendErrorQueueEmailNotificationFuture(List<String> sendList) {
-        sendErrorQueueEmailNotification(null, sendList);
+        ERR_Notifier notifierService = ERR_Notifier.getInstance();
```
[the remark this answers is not in the document]
Comment (iskander-sfdc) [reply force-app/main/default/classes/ERR_Notifier.cls:199]: Removed this parameter comment

File: force-app/main/default/classes/ERR_Notifier_TEST.cls (hunk 1 above)
Comment (iskander-sfdc) [reply force-app/main/default/classes/ERR_Notifier_TEST.cls:114]: I separated this test into 2 different test methods. This was needed to make sure it meets the filter condition which is based on Datetime.now()

File: force-app/main/default/classes/ERR_Notifier_TEST.cls (hunk 2 above)
Comment (iskander-sfdc) [reply force-app/main/default/classes/ERR_Notifier_TEST.cls:108]: Makes sense. Added 2 different test methods to test record count and age limits.

File: force-app/main/default/classes/ERR_Notifier_TEST.cls (hunk 3 above)
Comment (iskander-sfdc) [reply force-app/main/default/classes/ERR_Notifier_TEST.cls:288]: These helper methods are specific for this unit test so I decided to have them here for now

File: force-app/main/default/classes/ERR_Notifier.cls (hunk 4 above)
Comment (iskander-sfdc) [reply force-app/main/default/classes/ERR_Notifier.cls:130]: Right, this one is not needed, removed.

File: force-app/main/default/classes/ERR_Notifier.cls (hunk 5 above)
[the remark this answers is not in the document]
Comment (kg345) [reply force-app/main/default/classes/ERR_Notifier.cls:46]: Thanks for explaining - that makes sense

**Review by kg345 — APPROVED** [review]
Looks good! Added a new [test suite](https://gus.lightning.force.com/lightning/r/ADM_Test_Suite__c/a81EE000000ZXKTYA4/view) and scenarios for the apex tests added in this story.

**Pull request merged**